### PR TITLE
8320232: Cells duplicated when table collapsed and expanded

### PR DESCRIPTION
**Pull request by Maran23**
This PR fixes a bug where a `TableView` inside a `TitledPane` may 'duplicate' the cells. This actually has nothing to do with the `TitledPane`, but it triggered the bug easily due to its nature to change the size of his content when collapsing.

The expansion change of a `TitledPane` triggered an event where the underlying `VirtualFlow` was adding cells to the pile (for reuse) and later cleaning them all up without resetting the index to -1. 
This led to a bug where two cells had the same index and therefore received an edit event, although just one should (and is visible, the other cell has no parent -> orphan node).

The fix is to always reset the index to -1. This was already done before, just not everywhere, for all cells.
So before we clear the pile (or cells), we always reset the index to -1.
Added a bunch of tests, some pass before and after.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8320232](https://bugs.openjdk.org/browse/JDK-8320232): Cells duplicated when table collapsed and expanded (**Bug** - P3)

### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1521/head:pull/1521` \
`$ git checkout pull/1521`

Update a local copy of the PR: \
`$ git checkout pull/1521` \
`$ git pull https://git.openjdk.org/jfx.git pull/1521/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1521`

View PR using the GUI difftool: \
`$ git pr show -t 1521`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1521.diff">https://git.openjdk.org/jfx/pull/1521.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1521#issuecomment-2256668585)

### CODE REVIEW

**Comment by andy-goryachev-oracle** [comment]
Does this change affect `TreeView`?

**Comment by Maran23** [comment]
> Does this change affect `TreeView`?

All components that use the `VirtualFlow`,but it might be more noticeable for some than others. In fact, I first noticed this behavior  in `Dialog`s. I will add a reproducer for that one ticket as well soon, so it contains more information for people that may dig the backlog or git history in the future.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java [hunk 1]
```diff
@@ -805,6 +805,9 @@ public final BooleanProperty verticalProperty() {
         if (vertical == null) {
             vertical = new BooleanPropertyBase(true) {
                 @Override protected void invalidated() {
+                    resetIndex(cells);
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:808]: minor: [cells|pile]clear() can be combined with resetIndex(), something like

```
    private void clear(ArrayLinkedList<T> cells) {
        for (T cell : cells) {
            cell.updateIndex(-1);
        }
        cells.clear();
    }
```

as these operations seem to be called in pairs

**Review by andy-goryachev-oracle — COMMENTED** [review]
This is a definitely the right fix.

The TitledPane reproducer fails without and works correctly with the change.  I don't see the duplicate startEdit() with the Dialog reproducer, please clarify the steps to reproduce and what is expected/observed.

Left a couple of minor comments, will re-approve if you choose to make changes.

**Review by andy-goryachev-oracle — APPROVED** [review]
As with everything involving VirtualFlow, I'd like a second reviewer.  Maybe @johanvos ?

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java (hunk 1 above)
Comment (Maran23) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:808]: Unfortunately not in the `needsRebuildCells` if condition, where we only reset the index and add all cells to the pile

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java:808]: you are right.

**Comment by andy-goryachev-oracle** [comment]
I see the issue with the dialog reproducer on Windows 11.  macOS does not exhibit the issue with the dialog one, but does have an issue with the titled pane reproducer.

This change fixes the bug on macOS and Windows 11.

**Review by kevinrushforth — APPROVED** [review]
Looks good. I confirm that 3 of the newly-added test methods fail without the fix and pass with the fix.

I agree with Andy about the unrelated change. I'll reapprove if you fix it.

**Comment by Maran23** [comment]
/integrate

I agree the minor change is unrelated, but since it is that small for a test I had written in the past, I decided to keep it.

**Pull request closed without merging**